src/cargo/sources/git/utils.rs

```diff
@@ -565,7 +565,14 @@ pub fn fetch(repo: &git2::Repository, url: &str,
         let mut opts = git2::FetchOptions::new();
         opts.remote_callbacks(cb)
             .download_tags(git2::AutotagOption::All);
-        try!(remote.fetch(&[refspec], Some(&mut opts), None));
-        Ok(())
+
+        match network::with_retry(cargo_config, ||{
+            remote.fetch(&[refspec], Some(&mut opts), None)
+        }){
+            Ok(_) => Ok(()),
+            Err(e) =>{
+                Err(git2::Error::from_str(&format!("{}", e)))
```

Comment:
with_authentication is a closure that wraps the with_retry. this commit https://github.com/rust-lang/cargo/commit/537f7185422e8839986ae2ec29a341009b48e418#diff-7c9ac7738c4bef1573b9bbc2945334c8R391 changed the return of the with_authentication closure  from CargoResult to Result<T, git2::Error>. There is code for errors like git2::Error to convert to Box<CargoError> but Box<CargoError> does not convert to  Result<git2::Error>. If I am understanding the way errors are setup.